Fix external toolchain

francois-berder-imgtec opened the pull request:
this connects #33 

Some feeds rely upon CONFIG_USE_MUSL, CONFIG_USE_UCLIBC, CONFIG_USE_GLIBC
being set for compiling correctly packages. However, if using an external
toolchain, these variables are not set, resulting in build failures.
This commit ensures that at least one of these variable is set. It also
sets the appropriate value of TOOLCHAIN_LIBC depending on which libc
implementation is selected.

Signed-off-by: Francois Berder <francois.berder@imgtec.com>

francois-berder committed:
toolchain: Sets libc related variables if using external toolchain

Some feeds rely upon CONFIG_USE_MUSL, CONFIG_USE_UCLIBC, CONFIG_USE_GLIBC
being set for compiling correctly packages. However, if using an external
toolchain, these variables are not set, resulting in build failures.
This commit ensures that at least one of these variable is set. It also
sets the appropriate value of TOOLCHAIN_LIBC depending on which libc
implementation is selected.

Signed-off-by: Francois Berder <francois.berder@imgtec.com>

Ham22 commented:
The CI system won't be able to test this properly yet. So has this been verified use external toolchain options for both:
- The img official toolchain with glibc
- A pre-built toolchain produced by openwrt with musl

ghost commented:
In make menuconfig, Base System -> libc configuration, the pattern must be changed such that it can find libc.so and ld-mipsel-musl-sf.so.1

ghost commented:
@Ham22 I only tried using the pre-built toolchain produced by openwrt with musl.

Ham22 commented:
To get this working I have to set the libc configuration as you mentioned as below. The change required is pretty generic so I don't see why it couldn't just go in to the code defaults?

`./lib/ld{-*.so,*.so.*} ./lib/lib{anl,c,cidn,crypt,dl,m,nsl,nss_dns,nss_files,resolv,util}{.so,-*.so,.so.*}`

francois-berder committed:
toolchain: Fix default value of LIBC_FILE_SPEC

The default value of LIBC_FILE_SPEC does not work if using musl
because it does not match ld-musl-mipsel-sf.so.1 and libc.so.

Signed-off-by: Francois Berder <francois.berder@imgtec.com>

Ham22 commented:
Tested with musl pre-built, worked with below config added:
CONFIG_DEVEL=y
CONFIG_EXTERNAL_TOOLCHAIN=y
CONFIG_TARGET_NAME="mipsel-openwrt-linux-musl"
CONFIG_TOOLCHAIN_PREFIX="mipsel-openwrt-linux-musl-"
CONFIG_TOOLCHAIN_ROOT="/home/build/toolchain-mipsel_mips32_musl-1.1.15"
CONFIG_EXTERNAL_LIBC_USE_MUSL=y
